tools/stronghold/src/api/compatibility.py

```diff
@@ -150,43 +142,49 @@ def _check_by_position(
     func: str, before: api.Parameters, after: api.Parameters
 ) -> Iterable[Violation]:
     """Checks for violations among the positional parameters."""
-    before_params = [param for param in before.parameters if not param.keyword]
-    after_params = [param for param in after.parameters if not param.keyword]
 
-    if before_params == after_params:
-        return []
+    before_params = [param for param in before.parameters if param.position is not None]
+    after_params = [param for param in after.parameters if param.position is not None]
+
+    before_param_names = [param.name for param in before_params]
+    after_param_names = [param.name for param in after_params]
+
+    if before_param_names == after_param_names:
+        return
+
+    if set(before_param_names) == set(after_param_names):
+        yield Violation(
+            func,
+            'positional parameters were reordered',
+            line=after.line,
+        )
+        return
 
-    for i, (before_param, after_param) in enumerate(
-        itertools.zip_longest(before_params, after_params)
-    ):
-        assert before_param is None or before_param.position == i
-        assert after_param is None or after_param.position == i
-
-        if before_param is None:
-            assert after_param is not None
-            if after_param.required:
-                yield Violation(
-                    func,
-                    f'{after_param.name} was added and is required',
-                    line=after.line,
-                )
-        elif after_param is None:
-            assert before_param is not None
-            yield Violation(func, f'{before_param.name} was removed', line=after.line)
-        else:
-            assert before_param is not None
-            assert after_param is not None
-            if before_param.name != after_param.name:
-                yield Violation(
-                    func,
-                    f'{before_param.name} was renamed to {after_param.name}',
-                    line=after.line,
-                )
-                continue
-            if not before_param.required and after_param.required:
-                yield Violation(
-                    func, f'{before_param.name} became required', line=after.line
-                )
+    matcher = difflib.SequenceMatcher(a=before_param_names, b=after_param_names)
```

Comment:
I didn't realize you could use difflib like this 😮

Reply:
me neither, but it's pretty awesome, right?

I realized yesterday that diffing was basically what I wanted to do, and I think I looked at difflib with 5% optimism and the rest just for inspiration, but it paid off!